modules/raft/src/integrationTest/java/org/apache/ignite/raft/server/ItJraftCounterServerTest.java

```diff
@@ -709,6 +733,85 @@ public void testTimerThreadsCount() {
                 "All timer threads: " + timerThreads.toString());
     }
 
+    /**
+     * The test shows that all committed updates are applied after a RAFT group restart automatically.
+     * Actual data be available to read from state storage (not a state machine) directly just after the RAFT node started.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testApplyUpdatesOutOfSnapshot() throws Exception {
+        HashMap<Integer, AtomicInteger> counters = new HashMap<>(3);
+        HashMap<Path, Integer> snapshotDataStorage = new HashMap<>(3);
+        HashMap<String, SnapshotMeta> snapshotMetaStorage = new HashMap<>(3);
+
+        for (int i = 0; i < 3; i++) {
+            var counter = counters.computeIfAbsent(i, (key) -> new AtomicInteger());
+
+            startServer(i, raftServer -> {
+                raftServer.startRaftGroup("test_raft_group", new UpdateCountRaftListener(counter, snapshotDataStorage), INITIAL_CONF,
+                        defaults().snapshotStorageFactory(new SnapshotInMemoryStorageFactory(snapshotMetaStorage)));
+            }, opts -> {
+            });
+        }
+
+        var raftClient = startClient("test_raft_group");
+
+        raftClient.refreshMembers(true).get();
+        var peers = raftClient.peers();
+
+        raftClient.run(new TestWriteCommand());
+
+        assertTrue(TestUtils.waitForCondition(() -> counters.get(0).get() == 1, 10_000));
+
+        raftClient.snapshot(peers.get(0)).get();
+
+        raftClient.run(new TestWriteCommand());
+
+        assertTrue(TestUtils.waitForCondition(() -> counters.get(1).get() == 2, 10_000));
+
+        raftClient.snapshot(peers.get(1)).get();
+
+        raftClient.run(new TestWriteCommand());
+
+        for (AtomicInteger counter : counters.values()) {
+            assertTrue(TestUtils.waitForCondition(() -> counter.get() == 3, 10_000));
+        }
+
+        shutdownCluster();
+
+        Path peer0SnapPath = dataPath.resolve("node" + 0)
+                .resolve("test_raft_group" + "_" + peers.get(0).address().toString().replace(':', '_')).resolve("snapshot");
```

Comment:
Is there a helper method somewhere to mask special characters?

Reply:
I did not find a method, but I can write it int the test class.